opal/corelib/array.rb

```diff
@@ -2240,7 +2251,7 @@ def uniq!(&block)
   def unshift(*objects)
     %x{
       for (var i = objects.length - 1; i >= 0; i--) {
-        self.unshift(objects[i]);
+        unshiftOne(self, objects[i]);
       }
```

Comment:
@elia we can also provide a more optimized implementation when there's a list of objects to iterate on the array (self) only once.

Reply:
Untested code:

```js
function unshiftAll(list, elems) {
  var listLength = list.length
  var elemsLength = elems.length
  var index = listLength - elemsLength
  for (var i = 0; i < elemsLength; i++) {
    list.push(list[index + i])
  }
  var len = listLength - 1 
  while (len - elemsLength >= 0) {
    list[len] = list[len - elemsLength]
    len--
  }
  for (var i = 0; i < elemsLength; i++) {
    list[i] = elems[i]
  }
  return list.length
}
```

Reply:
I played a bit with this code and this version seems pretty good, can you please try it on different versions?

```rb
    function unshiftAll(list, elems) {
      for (var i = (list.length+elems.length)-1; i >= 0; i--) {
        if (i >= elems.length)
          list[i] = list[i-elems.length]
        else
          list[i] = elems[i]
      }
      return list.length
    }
```

probably the lengths can be cached into vars as you did above

Reply:
> probably the lengths can be cached into vars as you did above

We cannot cache `list.length` because we are adding elements in the loop so we need to call it at the end (just before the return).